python/ray/data/aggregate.py

```diff
@@ -92,42 +91,119 @@ def accumulate_block(a: AggType, block: Block) -> AggType:
         if finalize is None:
             finalize = lambda a: a  # noqa: E731
 
+        self.name = name
         self.init = init
         self.merge = merge
-        self.name = name
         self.accumulate_block = accumulate_block
         self.finalize = finalize
 
-    def _validate(self, schema: Optional[Union[type, "pa.lib.Schema"]]) -> None:
+    def _validate(self, schema: Optional["Schema"]) -> None:
         """Raise an error if this cannot be applied to the given schema."""
         pass
 
 
-class _AggregateOnKeyBase(AggregateFn):
-    def _set_key_fn(self, on: str):
-        self._key_fn = on
+@PublicAPI(stability="alpha")
+class AggregateFnV2(AggregateFn, Generic[AggType]):
```

Comment:
I think it'd be cleaner to not let AggregateFnV2 inherit AggregateFn. 
They should be 2 independent  classes. 
The `Dataset.aggregate` can take both V1 and V2.
If V1 is passed in, convert the V1 object to V2. 
And then all implementations should be based on V2. 
Updating the implementations may require non-trivial changes. We can do that in the next PR if the changes are too big.

Reply:
> I think it'd be cleaner to not let AggregateFnV2 inherit AggregateFn.
They should be 2 independent classes.

Yes, on the same page. Keep in mind this current arrangement is temporary until we completely drop V1